tools/network/dnssec/trustedzone.go

```diff
@@ -0,0 +1,208 @@
+// Copyright (C) 2019-2020 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package dnssec
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/miekg/dns"
+)
+
+type trustedZone struct {
+	name  string
+	zsk   map[uint16]dns.DNSKEY // zone signing keys
+	ksk   map[uint16]dns.DNSKEY // key signing keys
+	rrSig map[uint16]dns.RRSIG  // DNSKEY RR signature(s), used for validity
+	dss   []dns.DS              // DS or root anchor authenticating this zone in DS format
+}
+
+func (tz *trustedZone) isExpired(t time.Time) bool {
+	for _, sig := range tz.rrSig {
+		if !sig.ValidityPeriod(t) {
+			return true
+		}
+	}
+	return false
+}
+
+// ensure at least one of the keytags in ZSK of this zone
+func (tz *trustedZone) checkKeys(keytags []uint16) bool {
+	for _, kt := range keytags {
+		if _, ok := tz.zsk[kt]; ok {
+			return true
+		}
+	}
+	return false
+}
+
+// verifyDS checks DS RRSIG using this zone ZSK
+// cacheOutdated indicates no ZSK in this zone to verify the provided RRSIG
+func (tz *trustedZone) verifyDS(rrSet []dns.RR, rrSig []dns.RRSIG, t time.Time) (cacheOutdated bool, err error) {
+	// parent zone's DNSKEY RRSET is cached but DS is just came from the network
+	// and might be signed by newer key
+	kt := make([]uint16, 0, len(rrSig))
+	for _, sig := range rrSig {
+		kt = append(kt, sig.KeyTag)
+	}
+	// so if no such keys in a cache return cache outdated that must force this zone update
+	if !tz.checkKeys(kt) {
+		cacheOutdated = true
+		return
+	}
+
+	// at least one signature must be valid
+	verifiedSig := verifyRRSig(rrSet, rrSig, t, tz.zsk)
+	if len(verifiedSig) == 0 {
+		requestedZone := (rrSig)[0].Hdr.Name
+		err = fmt.Errorf("DS signature verification failed for %s", requestedZone)
```

Comment:
format the requestedZone within single quotes.

Reply:
done